File: client/lib/constants/message_types.dart
/// 三端共享的 WS 消息类型常量定义。
///
/// server / agent / client 三端必须保持同步。
/// 新增消息类型时，必须三端同时更新。
///
/// PLAINTEXT_MSG_TYPES: 不加密传输的消息类型集合（协议握手/心跳）。
/// 不在 PLAINTEXT_MSG_TYPES 中的消息必须加密传输。
class MessageType {
  MessageType._();

  // ---- 协议握手 / 心跳（PLAINTEXT） ----
  static const String auth = 'auth';
  static const String connected = 'connected';
  static const String ping = 'ping';
  static const String pong = 'pong';

  // ---- 终端数据 ----
  static const String data = 'data';
  static const String output = 'output';

  // ---- 终端控制 ----
  static const String resize = 'resize';
  static const String createTerminal = 'create_terminal';
  static const String closeTerminal = 'close_terminal';
  static const String terminalCreated = 'terminal_created';
  static const String terminalClosed = 'terminal_closed';
  static const String terminalsChanged = 'terminals_changed';

  // ---- 快照 ----
  static const String snapshot = 'snapshot';
  static const String snapshotStart = 'snapshot_start';
  static const String snapshotChunk = 'snapshot_chunk';
  static const String snapshotComplete = 'snapshot_complete';
  static const String snapshotRequest = 'snapshot_request';
  static const String snapshotData = 'snapshot_data';

  // ---- 执行命令 ----
  static const String executeCommand = 'execute_command';
  static const String executeCommandResult = 'execute_command_result';

  // ---- 知识库 / 工具 ----
  static const String lookupKnowledge = 'lookup_knowledge';
  static const String lookupKnowledgeResult = 'lookup_knowledge_result';
  static const String toolCall = 'tool_call';
  static const String toolResult = 'tool_result';
  static const String toolCatalogSnapshot = 'tool_catalog_snapshot';

  // ---- Agent 元数据 ----
  static const String agentMetadata = 'agent_metadata';

  // ---- 在线状态 ----
  static const String presence = 'presence';

  // ---- 连接管理 ----
  static const String deviceKicked = 'device_kicked';
  static const String error = 'error';
}

/// 不加密的控制消息类型（协议握手/心跳）
const Set<String> plaintextMsgTypes = {
  MessageType.auth,
  MessageType.connected,
  MessageType.ping,
  MessageType.pong,
};

/// 判断消息类型是否需要加密
bool shouldEncrypt(String msgType) {
  return !plaintextMsgTypes.contains(msgType);
}

File: server/app/infra/message_types.py
"""
三端共享的 WS 消息类型常量定义。

server / agent / client 三端必须保持同步。
新增消息类型时，必须三端同时更新。

PLAINTEXT_MSG_TYPES: 不加密传输的消息类型集合（协议握手/心跳）。
不在 PLAINTEXT_MSG_TYPES 中的消息必须加密传输。
"""
from enum import StrEnum


class MessageType(StrEnum):
    """WebSocket 消息类型枚举。"""

    # ---- 协议握手 / 心跳（PLAINTEXT） ----
    AUTH = "auth"
    CONNECTED = "connected"
    PING = "ping"
    PONG = "pong"

    # ---- 终端数据 ----
    DATA = "data"
    OUTPUT = "output"

    # ---- 终端控制 ----
    RESIZE = "resize"
    CREATE_TERMINAL = "create_terminal"
    CLOSE_TERMINAL = "close_terminal"
    TERMINAL_CREATED = "terminal_created"
    TERMINAL_CLOSED = "terminal_closed"
    TERMINALS_CHANGED = "terminals_changed"

    # ---- 快照 ----
    SNAPSHOT = "snapshot"
    SNAPSHOT_START = "snapshot_start"
    SNAPSHOT_CHUNK = "snapshot_chunk"
    SNAPSHOT_COMPLETE = "snapshot_complete"
    SNAPSHOT_REQUEST = "snapshot_request"
    SNAPSHOT_DATA = "snapshot_data"

    # ---- 执行命令 ----
    EXECUTE_COMMAND = "execute_command"
    EXECUTE_COMMAND_RESULT = "execute_command_result"

    # ---- 知识库 / 工具 ----
    LOOKUP_KNOWLEDGE = "lookup_knowledge"
    LOOKUP_KNOWLEDGE_RESULT = "lookup_knowledge_result"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"
    TOOL_CATALOG_SNAPSHOT = "tool_catalog_snapshot"

    # ---- Agent 元数据 ----
    AGENT_METADATA = "agent_metadata"

    # ---- 在线状态 ----
    PRESENCE = "presence"

    # ---- 连接管理 ----
    DEVICE_KICKED = "device_kicked"
    ERROR = "error"


# 不加密的控制消息类型（协议握手/心跳）
PLAINTEXT_MSG_TYPES: frozenset[str] = frozenset({
    MessageType.AUTH,
    MessageType.CONNECTED,
    MessageType.PING,
    MessageType.PONG,
})


# ---- WebSocket 关闭码常量 ----
# 集中定义，避免散落在各 ws 模块中。
# 4xxx 为应用自定义码。

class WSCloseCode:
    """WebSocket 关闭码常量。"""
    # 协议标准码
    NORMAL = 1000        # 正常关闭
    ABNORMAL = 1006      # 无 close frame（网络中断）
    POLICY_VIOLATION = 1008  # 心跳超时

    # 应用自定义码 — 认证/协议
    AUTH_TIMEOUT = 4002      # 认证超时
    PROTOCOL_ERROR = 4003    # 协议错误（缺少加密/Token 缺字段）
    INVALID_MESSAGE = 4004   # 无效消息（JSON 错误/非 auth 消息）

    # 应用自定义码 — 业务
    SESSION_CONFLICT = 4009  # Session/terminal 冲突
    DEVICE_REPLACED = 4011   # 设备被新连接替换

    # 应用自定义码 — 客户端错误
    INVALID_VIEW = 4400      # 无效 view_type

    # 应用自定义码 — 服务端错误
    INTERNAL_ERROR = 4500    # 服务端内部错误
    NETWORK_LOST = 4501      # Agent 网络丢失
    TOO_MANY_CLIENTS = 4503  # 客户端连接数超限
